python/ray/tests/test_multiprocessing.py

```diff
@@ -0,0 +1,502 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+import pytest
+import tempfile
+import time
+import random
+import subprocess
+from collections import defaultdict
+import queue
+
+import ray
+from ray.experimental.multiprocessing import Pool, TimeoutError
+
+
+@pytest.fixture
+def cleanup_only():
+    yield None
+    ray.shutdown()
+    subprocess.check_output(["ray", "stop"])
+    if "RAY_ADDRESS" in os.environ:
+        del os.environ["RAY_ADDRESS"]
+
+
+@pytest.fixture
+def pool():
+    pool = Pool(processes=1)
+    yield pool
+    pool.terminate()
+    ray.shutdown()
+
+
+@pytest.fixture
+def pool_4_processes():
+    pool = Pool(processes=4)
+    yield pool
+    pool.terminate()
+    ray.shutdown()
+
+
+def test_initialize_ray(cleanup_only):
+    def getpid(args):
+        import os
+        return os.getpid()
+
+    def check_pool_size(pool, size):
+        args = [tuple() for _ in range(size)]
+        assert len(set(pool.map(getpid, args))) == size
+
+    # Check that starting a pool starts ray if not initialized.
+    pool = Pool(processes=2)
+    assert ray.is_initialized()
+    assert int(ray.state.cluster_resources()["CPU"]) == 2
+    check_pool_size(pool, 2)
+    ray.shutdown()
+
+    # Check that starting a pool doesn't affect ray if there is a local
+    # ray cluster running.
+    ray.init(num_cpus=3)
+    assert ray.is_initialized()
+    pool = Pool(processes=2)
+    assert int(ray.state.cluster_resources()["CPU"]) == 3
+    check_pool_size(pool, 2)
+    ray.shutdown()
+
+    # Check that trying to start a pool on an existing ray cluster throws an
+    # error if there aren't enough CPUs for the number of processes.
+    ray.init(num_cpus=1)
```

Comment:
Not regarding your PR, but just wanted to leave this as a note so we don't forget for the future: we should consider a `with`-based API for starting and shutting down Ray instead of manually doing the two separately every time.

Reply:
`ray.shutdown()` is actually basically only used for testing AFAIK. We should definitely improve the `ray.init()` API though.